README.md

```diff
@@ -237,18 +465,107 @@ sys_hook.enable()
 
 print(sys_hook.hit_count)
 ```
-Note: there can be at most one hook for each syscall.
+Note: There can be at most one user-defined hook for each syscall. \
+The pretty print function (described below) is not considered a user-defined hook. Therefore, it is possible to hook/hijack and pretty print the same syscall simultaneously. \
+If a new hook is defined for a syscall that is already hooked or hijacked, the new hook replaces the old one, and a warning is shown.
+
+For example, the following code
+
+```python
+def on_enter_open_1(d: ThreadContext, syscall_number: int):
+    print("entering open 1")
+    d.syscall_arg0 = 0x1
+
+def on_exit_open_1(d: ThreadContext, syscall_number: int):
+    print("exiting open 1")
+    d.syscall_return = 0x0
+
+def on_enter_open_2(d: ThreadContext, syscall_number: int):
+    print("entering open 2")
+    d.syscall_arg0 = 0x4
+
+def on_exit_open_2(d: ThreadContext, syscall_number: int):
+    print("exiting open 2")
+    d.syscall_return = 0xffffffffffffffff
+
+sys_hook_1 = d.hook_syscall(syscall="open", on_enter=on_enter_open_1, on_exit=on_exit_open_1)
+
+sys_hook_2 = d.hook_syscall(syscall="open", on_enter=on_enter_open_2, on_exit=on_exit_open_2)
+```
+will override `sys_hook_1` with `sys_hook_2` showing the following warning
+
+![alt text](media/syscall_hook_override_warning.png)
+
+## Syscall Hijacking
+libdebug also facilitates easy hijacking of syscalls, as shown in the following example:
+```py
+from libdebug import debugger
+
+d = debugger("/usr/bin/ls")
+d.run()
+
+hook = d.hijack_syscall("read", "write")
+# hook = d.hijack_syscall("read", 0x1)
+# hook = d.hijack_syscall(0x0, 0x1)
+
+d.cont()
+[...]
+```
+#### Syscall Hijacking Loop Detection
+During execution, libdebug checks for loops in syscall hijacking and raises an exception if infinite loops are detected. \
+For example, the following code
+```py
+from libdebug import debugger
+
+d = debugger("/usr/bin/ls")
+d.run()
+
+hook = d.hijack_syscall("read", "write")
+hook = d.hijack_syscall("write", "read")
+
+d.cont()
+[...]
+```
+raises the following execption
+
+![alt text](media/syscall_hijacking_loop_detection.png)
+
+#### Hook on hijack
+After a syscall is hijacked, the user can choose whether to execute the hook or hijacking installed in the newly executed syscall, if any. \
+
+For example, if we hijack the syscall `read` with the syscall `write`, and the `write` is also hooked or hijacked, `hook_hijack` allows us to choose whether to execute the hook/hijack installed on the `write` when the read becomes a `write`. This helps reduce loops during hijacking and behaviors that are difficult to track
+
+The syscall hijacking loop detection takes this choice into account.
+```py
+from libdebug import debugger
+
+def on_enter_write(d: ThreadContext, syscall_number: int):
```

Comment:
I just noticed that this is inconsistent with the Breakpoint API, where in the callback we pass the Breakpoint directly.